quickjs.c

```diff
@@ -12009,6 +12009,20 @@ int JS_ToBigInt64(JSContext *ctx, int64_t *pres, JSValue val)
     return JS_ToBigInt64Free(ctx, pres, js_dup(val));
 }
 
+int JS_ToBigUint64(JSContext *ctx, uint64_t *pres, JSValue val)
+{
+    bf_t a_s, *a;
+
+    a = JS_ToBigIntFree(ctx, &a_s, js_dup(val));
+    if (!a) {
+        *pres = 0;
+        return -1;
+    }
+    bf_get_uint64(pres, a);
+    JS_FreeBigInt(ctx, a, &a_s);
+    return 0;
+}
+
```

Comment:
What is the expected behavior of `JS_ToBigUint64` for values larger or equal to 2<sup>64</sup> and for negative values?  As a matter of fact, the behavior of `JS_ToBigInt64` for values outside the range of `int64_t` is unclear to me too.

Reply:
Well, I'd say it'd defined as extracting the value in the `val` BigInt JS object, right? Whatever a JS BigInt objectg can represent, that's what what you'll get here.

Reply:
Yes, but what if the BigInt value is out of range?  Should these functions return the value modulo 2<sup>64</sup>  as specified in ECMA `7.1.7 ToUint32` for example or should they clamp the value and return `UINT64_MAX` for values beyond and `0` for negative values.
I much prefer the former (modulo) and it seems to be what the implementation of `JS_ToBigInt64` does, but your implementation of `JS_ToBigUint64` does the second, which is inconsistent.
The implementation is somewhat tricky because the current implementation of BigInt uses sign+magnitude and a mantissa/exponent representation, but this will change when Fabrice completes his reimplementation of BigInt using a more standard approach and 2's complement representation.
What is the current use case for `JS_ToBigUint64` ?

Reply:
I see. Looks like libbf would need to be extended then, since now all it does is set it to `UINT64_MAX`. It's also currently unused by quickjs itself, so there is that.

Regarding motivation / use case, I have none, personally. It was requested in https://github.com/quickjs-ng/quickjs/issues/376 and it looked like low hanging fruit, but maybe it isn't after all 😅

Reply:
Looking again, there are 2 abstract operations defined by ECMA with the corresponding semantics:
- **7.1.15 `ToBigInt64`**
- **7.1.16 `ToBigUint64`**
Both compute the 64-bit value modulo 2<sup>64</sup>, which means they can be implemented using the same internal function and a simple cast on the pointer argument:
```
int JS_ToBigUint64(JSContext *ctx, uint64_t *pres, JSValue val)
{
    return JS_ToBigInt64Free(ctx, (int64_t *)pres, js_dup(val));
}
```

Reply:
I see! Let's go with the simple one then!